### PR TITLE
fix/5075

### LINKED ISSUE #5075: API: delete of a node is possible even with a locationlog entry opened
**Describe the bug**
From web admin, it's possible to delete a node even if this node has a locationlog entry opened. That's not the case with `pfcmd node delete XX:XX:XX:XX:XX:XX`.

From my point of view, we should have same behavior between two tools and I prefer `pfcmd` approach.

**To Reproduce**
1. Generate a RADIUS request (to create a locationlog entry):
``` 
# cat >> /opt/tls.config << EOF
network={
ssid="CORPORATE SSID"
key_mgmt=WPA-EAP
eap=TLS
identity="inverse"
ca_cert="/usr/local/pf/raddb/certs/ca.pem"
client_cert="/usr/local/pf/raddb/certs/client.pem"
private_key="/usr/local/pf/raddb/certs/client.key"
private_key_passwd="whatever"
eapol_flags=3
}
EOF

# eapol_test -c /opt/tls.config -s testing123 -a MGMT_IP -A 127.0.0.1 -N 61:d:15 -M 11:11:11:11:11:11 -N 4:x:0xc0a80101
```
2. Click on node added
3. Click on Delete

=> Your node is removed from `node` table but its location stays in place in `locationlog` table.

**Expected behavior**
It should not possible to delete a node if a locationlog entry is opened.

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Node deletion in consistent between the the API and pf::node::node_delete

Impacts
=======
API

NEWS file entries
=======================
Enhancements
-------------
* Node deletion in consistent between the the API and pf::node::node_delete

Issue
=====
fixes #5075

Delete branch after merge
=========================
NO

Checklist
=========
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

File: lib/pf/node.pm [hunk 1]
```diff
@@ -156,6 +156,16 @@ sub node_view_reg_pid {
...
+sub _can_delete {
+    my ($mac) = @_;
+    if ( defined( pf::locationlog::locationlog_view_open_mac($mac) ) ) {
+        my $msg = "$mac has an open locationlog entry. Node deletion prohibited";
+        return (0, $msg);
+    }
+
+    return (1,'');
+}
+
```
Comment (nqb) [inline lib/pf/node.pm:168]: Should we not replace `0` and `1` by `$FALSE` and `$TRUE` ?

File: lib/pf/node.pm [hunk 2]
```diff
@@ -173,8 +183,9 @@ sub node_delete {
 
     require pf::locationlog;
     # TODO that limitation is arbitrary at best, we need to resolve that.
-    if ( defined( pf::locationlog::locationlog_view_open_mac($mac) ) ) {
-        $logger->warn("$mac has an open locationlog entry. Node deletion prohibited");
+    my ($results, $msg) = _can_delete($mac);
+    if (!$results) {
+        $logger->warn($msg);
```
Comment (nqb) [inline lib/pf/node.pm:186]: 1. Is TODO comment still necessary ?
1. I will prefer `$result` in place `$results` like in other files.

File: lib/pf/node.pm [hunk 3]
```diff
@@ -173,8 +183,9 @@ sub node_delete {
...
         return (0);
```
Comment (nqb) [inline lib/pf/node.pm]: Should we not replace `0` by `$FALSE` ?

**Review by nqb — CHANGES_REQUESTED** [review]

**Comment by nqb** [comment]
@jrouzierinverse,

Another thing I noticed. Now when you remove a node using `pfcmd`, you got two lines in logs:
```
Feb  3 09:50:06 pfcen7stable packetfence: WARN pfcmd.pl(29687): 11:11:11:11:11:11 has an open locationlog entry. Node deletion prohibited (pf::node::node_delete)
Feb  3 09:50:06 pfcen7stable packetfence: ERROR pfcmd.pl(29687): Cannot delete node 11:11:11:11:11:11 since there are some records in locationlog table indicating that this node might still be connected and active on the network  (pf::cmd::pf::node::action_delete)
```
and you get "Cannot delete node 11:11:11:11:11:11 .." on STDERR.

IMO, we should adjust `action_delete` function in `pf::cmd::pf::node` to:
- use message returns by API in place of this different message
- display message returns by API on STDERR

File: lib/pf/node.pm (hunk 2 above)
Comment (nqb) [reply lib/pf/node.pm:186]: @jrouzierinverse, what about the TODO item ?

Comment (jrouzierinverse) [reply lib/pf/node.pm:186]: I find $can_delete is more clear on the value is for.

**Comment by nqb** [comment]
@jrouzierinverse, what about my last comment ?

**Comment by nqb** [comment]
It works as expected.

**Pull request merged**